man/ffpaa.1.in
.de CW
. nop \f[C]\\$*\f[]
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpaa 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpaa \- This program generates an FFP vector of amino acid features.
.SH SYNOPSIS
.BI "ffpaa [" "OPTION" "] ... [" "FILE"  "] ..."
.SH DESCRIPTION
.PP
Given no options the default behavior of the program is to
generate a Feature Frequency Profile (FFP) using features of length
4.  By default amino acids are divided and counted using 11 equivalent
classes: (ST),(DE),(KQR),(IVLM),(FWY),C,G,A,N,H,P [Li et al. (2003) Peds, 16. 323].
For multicharacter classes the amino acid is reported in the ffp output by the first character 
in the series.  For example, (IVLM) is represented by I.
.SH OPTIONS
.TP
.BI "\-l " "LEN" ", --length=" "LEN"
.RI "Changes the default length of features to " "LEN" ". The default length is 4,"
with a maximum length allowed of 40.
You may overide this maximum length by setting (and exporting)
the environmental variable MAX_WORD_SIZE to a different value.
.TP
.BI "\-f " "FILE" ", --feature-list=" "FILE"
Changes the behavior of the program to read
.RI "a list of features from " "FILE" ". Features can"
be space or newline delimited.
.TP
.BI "\-w " "STR" ", --mask=" "STR"
.RI "Use character " "STR" " of length " "LEN" " which specifies"
a character mask for the features. For example "1111101111" means to ignore the 6th character of a feature.
.TP
.BI "\-z " "INT" ", --rand-mask=" "INT"
Create a random character mask which allows up to
.IR "INT" " mismatches.  No greater than " "LEN"
in mismatches is allowed. The character mask is
printed to standard error at the beginning of
execution.
.TP
.BI "\-s " "INT" ", --rand-seed=" "INT" 	
Seed the random number generator. Randomization 
.RB "operations like " "-z" " are fixed and  predictable across different runs."
By default the seed is (system time) * (process ID).
.TP
.BI "\-q, --quiet"
Run in quiet mode.  Suppresses printing to standard error.  This effects option
.BR "\-z" "."
.TP
.BI "\-d, --disable-classes"
Disable classing of amino acids. Use 20 symbol amino acid coding.
.TP
.B "\-m, --multiple"
.IR "FILE " "contains multiple FASTA sequences and an FFP is desired for each sequence."
.TP
.B "\-h, --help"
Display help message.
.TP
.B "\-v, --version"
Display version.
.PP
.SH EXAMPLES
The command:
.PP
ffpaa -l 3 file*.faa
.PP
Will produce a output of this form, using features of length 3:
.PP
.CODE AGK 2 ASS 3 AKF 4 ...
.br
.CODE AKK 3 AKS 2 AKF 5 ...
.PP
There will be one FFP line for each 
.CW .faa 
file supplied in the argument list and each will be printed in the order
specified on the command line.  Note, the above uses the order in which the
shell (i.e. bash, cshell, etc.) expands the file argument wildcard, 
.C *.faa.
In the even this is undesired, the format format below can also be used
to explicitly control the ordering:
.PP
.CODE ffpaa -l 4 file1.faa file2.faa file3.faa
.PP
A mask can be applied to feature counting to allow mismatches using the 
.B \-w 
command.  For example if a mismatch is to be allowed
at the second and fourth positions of a feature with length 
.IR "l" "=6, " "this can be done via:"
.PP
.CODE ffpaa -l 6 -w "101011" file*.fna
.PP
If mismatches are desired but no specific pattern is necessary
then one can use the
.B -z
option.
.PP
.CODE ffpaa -z 2 file1.faa file2.faa.  
.PP
This will produce a random mask of at least two mismatches (within a feature with
a default length of 4).  The mask
is printed to standard error. To save the mask for
later reference, you can direct standard error to a file.
For example:
.PP
.CODE ffpaa -z 3  file*.fna > vector 2> mask
.PP
The redirection
.CW >2
sends standard error to the file 
.CW mask.
.PP
.PP
To disable classing of amino acids, (i.e. make FFP consider
.RB "R and K different -- and therefore count each symbol separately) use the " "-d" " option"
.PP
.PP
The utilities in the FFP package can be used as fully
qualified filters and can be used to build pipelines.
The pipeline below will create a Phylip infile format distance
matrix, using the Jensen Shannon divergence:
.PP
.CODE ffpaa -l 7 file*.fna | ffpcol -a | ffprwn | ffpjsd -p taxa.txt > infile
.PP
Note, the desired taxa names must be specified in the argument to 
.BR "\-p" ","
see
.BR ffpjsd(1)
.
.PP
To disable classing of animo acids you should incorporate the 
.B \-d 
option into the above pipeline -- both in the call to 
.B ffpaa 
and 
.B ffpcol.
.PP
.CODE ffpaa -l 7 -d file*.fna | ffpcol -a -d | ffprwn ...
.PP
Bootstrap and jacknife resampling can be done with the
ffpboot program, see 
.BR ffpboot(1).
.PP
.SH FURTHER DIRECTIONS
Command line definable character classes.
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpry(1),
.BR ffprwn(1), 
.BR ffpjsd(1),
.BR ffpcol(1),
.BR ffpmerge(1),
.BR ffpboot(1),
.BR ffpvprof(1),
.BR ffpreprof(1),
.BR ffptree(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpaa.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <math.h>
#include <ctype.h>
#include <limits.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hashroll.h"
#include "mask.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "parse_features.h"
#include "../config.h"
#define DEFAULT_WORD_LENGTH 4 /**< Default Feature length if not given by opt -l */
#define CHAR_BUFFER_SIZE 10000 /**< Default buffer size for reading sequence files */


/** @Todo implement a variety of character classes 
 SE-B(14)	A, C, D, EQ, FY, G, H, IV, KR, LM, N, P, ST, W
 SE-B(10)	AST, C, DN, EQ, FY, G, HW, ILMV, KR, P
 SE-V(10)	AST, C, DEN, FY, G, H, ILMV, KQR, P, W
 Li-A(10)	AC, DE, FWY, G, HN, IV, KQR, LM, P, ST
 Li-B(10)	AST, C, DEQ, FWY, G, HN, IV, KR, LM, P
 Solis-D(10)	AM, C, DNS, EKQR, F, GP, HT, IV, LY, W
 Solis-G(10)	AEFIKLMQRVW, C, D, G, H, N, P, S, T, Y
 Murphy(10)	A, C, DENQ, FWY, G, H, ILMV, KR, P, ST
 SE-B(8)	AST, C, DHN, EKQR, FWY, G, ILMV, P
 SE-B(6)	AST, CP, DEHKNQR, FWY, G, ILMV
 Dayhoff(6)	AGPST, C, DENQ, FWY, HKR, ILMV

 These are from Edgar, R (2004, NAR, 32:380-385

 */


char PROG_NAME[FILENAME_MAX];

static void parseFile(HASH *, FILE *);
static void loopFeatureList(HASH * h, FILE * fp);
static void loopRaw(HASH * h, FILE * fp);

char *weightVector;  /**< A mask to allow mismatches in features */
int Length = DEFAULT_WORD_LENGTH;
				/**< Feature length to use if not specified by opt -l */
int Buffsize = CHAR_BUFFER_SIZE;/**< File input buffer, increase value for larger genomes. opt -b can be used to change this value */
int maxWordSize = MAX_WORD_SIZE;


bool wflag = false;	     /**<-w Supply a mismatch character mask */
char *wvalue = NULL; /**<-w ptr to mismatch character mask str */
bool zflag = false;	     /**<-z Create random mismatch character mask */
int zflagN = 0;	     /**<-z Maximum number of mismatches to allow */
bool sflag = false;	     /**<-s Fix random seed */
int sflagN = 0;	     /**<-s Random seed value */
bool qflag = false;	     /**<-q Force quiet operation */
bool fflag = false;	     /**<-f Restrict counting to a list of features*/
char *fvalue = NULL; /**<-f ptr to name of featmer mask file */
bool dflag = false;	     /**<-d disable classing of amino acids */
bool mflag = false;	     /**<-m option, FNA file contains multiple sequences */

char usage_str[] = "Usage: %s [OPTION] ... [FILE] ... \n\
This program generates an FFP vector of amino acid features\n\n\
Given no options the default behavior of the program is to\n\
generate a Feature Frequency Profile (FFP) using features of\n\
length 4.\n\
\t-l LEN, --length\n\
\t-f FILE, --feature-list\n\
\t-w STR, --mask\n\
\t-z INT, --rand-mask=INT\n\
\t-s INT, --rand-seed\n\
\t-q, --quiet\n\
\t-d, --disable-classes\n\
\t-m, --multiple\n\
\t-h, --help\n\
\t-v, --version\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";

int main(int argc, char **argv)
{
    int opt;
    FILE *fp;
    HASH h;

    int option_index = 0;

    static struct option long_options[] = {
	{"length", required_argument, 0, 'l'},
	{"disable-classes", no_argument, 0, 'd'},
	{"feature-list", required_argument, 0, 'f'},
	{"mask", required_argument, 0, 'w'},
	{"rand-mask", required_argument, 0, 'z'},
	{"rand-seed", required_argument, 0, 's'},
	{"quiet", no_argument, 0, 'q'},
	{"help", no_argument, 0, 'h'},
	{"multiple", no_argument, 0, 'm'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };

  initSignalHandlers();

    while ((opt = getopt_long(argc, argv, "l:dw:z:s:qf:h?mv",
			      long_options, &option_index)) != -1)

	switch (opt) {
	case 'l':
	    Length = atoi(optarg);
	    break;
	case 'w':
	    wflag = !wflag;
	    wvalue = optarg;
	    break;
	case 'z':
	    zflag = !zflag;
	    zflagN = atoi(optarg);
	    break;
	case 's':
	    sflag = !sflag;
	    sflagN = atoi(optarg);
	    break;
	case 'q':
	    qflag = !qflag;
	    break;
	case 'd':
	    dflag = !dflag;
	    break;
	case 'f':
	    fflag = !fflag;
	    fvalue = optarg;
	    break;
	case 'm':
	    mflag = !mflag;
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;
	}


    if ( getenv("MAX_WORD_SIZE") ) {
	maxWordSize=atoi(getenv("MAX_WORD_SIZE"));	    
    }	    

    if (Length > maxWordSize) 
	    fatal_msg("%d: Max Word size is : %d",Length,maxWordSize);

    if (sflag)
	srand((unsigned) sflagN);
    else
	srand((unsigned) time(NULL)*getpid());


    if (wflag) {
	weightVector = (char *) malloc(sizeof(char) * (Length + 1));
	strcpy(weightVector, wvalue);
	if (!qflag)
	    warn_msg("USING FEATURE MASK: %s\n", weightVector);

    }


    if (zflag) {
	while (zflagN > Length - 1)
	    zflagN--;
	weightVector = (char *) malloc(sizeof(char) * (Length + 1));
	randweight(weightVector, Length, zflagN);
	if (!qflag)
	    warn_msg("USING FEATURE MASK: %s\n", weightVector);
    }
    // Initialize the rolling hash
    // reverse is not applicable, therefore 0
    init(&h, (zflag || wflag), amino, !dflag, 0, Length);


    // If provided a feature list read it and store in hash

    if (fflag) 
	parseFeatureList(&h,fvalue,Length,amino);

// Must now process file arguments
    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
	    	fatal_msg("%s: %s\n", *argv, strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	parseFile(&h, fp);

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);

    //freeHash(&h);

    return EXIT_SUCCESS;
}




/**
 *
 *  Parses FASTA format FNA file
 *
 *  This function can read from stdin or from files.
 *  The speed of the function is affected by the size
 *  of the input buffer.  Multiple FFPs are printed
 *  out with the -m option for each >header
 *  in the file. Another flag which affects this function
 *  is -f
 *
 *  @param h The hash table to use.
 *  @param fp The FNA file to parse
 *  @return none
 *
 */


static void parseFile(HASH * h, FILE * fp)
{

// to eliminate a test w/n a tight loop
// First test here.
    if (fflag)
	loopFeatureList(h, fp);
    else
	loopRaw(h, fp);

}


// For finding only features w/n a feature list
static void loopFeatureList(HASH * h, FILE * fp)
{
    struct stat fattr;
    static size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s: File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    	if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    chkpushaaw(h, buf, nr,firstRecord);
	else
	    chkpushaa(h, buf, nr,firstRecord);

	firstRecord=false;
    }

    printFeatures(h);
    free(buf);
    freeHash(h);
}



// For finding all features

static void loopRaw(HASH * h, FILE * fp)
{
    struct stat fattr;
    static size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;


    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s: File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    	if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    pushaaw(h, buf, nr,firstRecord);
	else
	    pushaa(h, buf, nr,firstRecord);

	firstRecord=false;
    }

    printFeatures(h);
    freeHash(h);
    free(buf);
}

